nixos/modules/services/databases/postgresql.nix

```diff
@@ -168,6 +168,18 @@ in
                 }
               '';
             };
+            ensureRoles = mkOption {
+              type = types.listOf (types.enum [ "superuser" "create role" "create db" "replication" "bypass urls" ]);
+              default = [ ];
+              description = lib.mdDoc ''
+                A list of roles to grant to the user. Under the hood this uses the
+                [ALTER USER syntax](https://www.postgresql.org/docs/current/sql-alteruser.html) for each role in the list:
+                `ALTER USER user.name WITH role`
```

Comment:
Can clauses be removed? We don't seem to have enough knowledge to remove unmentioned roles.

The linked page lists negatives as well: `NOSUPERUSER`, etc. Could those be used for this purpose?
A better option representation would be `types.attrsOf types.bool`. (or even individual options, to restrict it to known-valid attribute names; not sure if that's better)

With this infrastructure in place, we can also provide an option (off by default, not to break configs) for fully declarative user permissions; a matter of conditionally assigning some `lib.mkDefault` defaults to those options, so that they cover all possible clauses.

Reply:
Great suggestion. I will try to find some time this weekend to do this, as well as adding tests.